Tests: Use Queue for process communication which replaces tmp files and use OS for port creation

MVrachev opened the pull request:
Fixes #1124, Fixes #1111, Fixes #1196 

**Description of the changes being introduced by the pull request**:

These changes can be summarized with the following bullets:
- Delegate generation of ports used for the tests to the OS
- Use thread-safe Queue for processes communication
instead of temporary files
- Remove all instances of port generation or hardcoded ports
- Make test_slow_retrieval.py fully conform with TestServerProcess
- Remove test_slow_retrieval expected failure test  (the test with mode_2)
- Add tests for `tests/utils.py`
- Remove all instances of `temp file` mentions in the comments regarding the instances of `TestServerProcess`. 

Delegate generation of ports used for the tests to the OS is much
better than if we manually generate them, because there is always
a chance that the port we have randomly pick turns out to be taken.
By giving 0 to the port argument we ask the OS to give us
an arbitrary unused port.

Use thread-safe Queue for processes communication instead of temporary
files became a necessity because of findings made by Jussi Kukkonen.
With the latest changes made in pr 1192 we were rapidly reading
from the temporary files and Jussi found that it happened rarely
the successful message "bind succeded..." to be corrupted.
It seems, this is a thread issue related to the thread redirecting
the subprocess stdout to the temp file and our thread rapidly
reading from the file.
By using a thread-safe Queue we eliminate this possibility.
For reference read:
#1196

Lastly, test_slow_retrieval.py and slow_retrieval.py were refactored.
Until now, slow_retrieval.py couldn't use the TestServerProcess class
from utils.py for a port generation because of a bug related to
httpd.handle_request().
Now, when we use httpd.serve_forever() we can refactor both of those
files and fully conform with TestServerProcess.

----------------------------------------------------------

This is the 3-rd pr which aims to fix the issues mentioned above.
The other two were closed:

1. [Add a retry mechanism on server startup failure and use TestServerProcess for port generation](https://github.com/theupdateframework/tuf/pull/1169)
2. [Delegate port generation for the tests to the OS ](https://github.com/theupdateframework/tuf/pull/1192)

The first one was closed because I realized that delegating the port generation to the OS was much better
then if we generated a port and then validate it.
The reasons are described here: https://github.com/theupdateframework/tuf/pull/1169#issuecomment-717908695

The second pr was closed because @jku found a rarely appearing thread bug.
You can read my comment about it here: https://github.com/theupdateframework/tuf/pull/1192#issuecomment-721723063

**PS:** Again, many thanks to @jku for reviewing and discussing those issues with me!
Those changes are a good example of what a community can achieve together!

**Please verify and check that the pull request fulfills the following
requirements**:
- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
Remove test_slow_retrieval expected failure test

Remove the test with mode 2 ('mode_2': During the download process,
the server blocks the download by sending just several characters
every few seconds.) from test_slow_retrieval.

This test is marked as "expected failure" with the purpose of
rewriting it one day, but slow retrievals have been removed from
the specification and soon it will be removed from the tuf
reference implementation as a whole.
That means that the chances of making this test useful are close
to 0 if not none.

The other test (with mode 1) in test_slow_retrieval is not removed.

For reference:
- https://github.com/theupdateframework/specification/pull/111
- https://github.com/theupdateframework/tuf/pull/1156

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Remove unused random module imports

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Remove sleep from test_slow_retrieval_attack.py

Now, after we can use wait_for_server and the retry mechanism
of TestServerProcess in utils.py we no longer need to use
sleep in this test file.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Tests: Queue replace tmp files, OS port creation

These changes can be summarized with the following bullets:
- Delegate generation of ports used for the tests to the OS
- Use thread-safe Queue for processes communication
instead of temporary files
- Remove all instances of port generation or hardcoded ports
- Make test_slow_retrieval.py fully conform with TestServerProcess

Delegate generation of ports used for the tests to the OS is much
better than if we manually generate them, because there is always
a chance that the port we have randomly pick turns out to be taken.
By giving 0 to the port argument we ask the OS to give us
an arbitrary unused port.

Use thread-safe Queue for processes communication instead of temporary
files became a necessity because of findings made by Jussi Kukkonen.
With the latest changes made in pr 1192 we were rapidly reading
from the temporary files and Jussi found that it happened rarely
the successful message "bind succeded..." to be corrupted.
It seems, this is a thread issue related to the thread redirecting
the subprocess stdout to the temp file and our thread rapidly
reading from the file.
By using a thread-safe Queue we eliminate this possibility.
For reference read:
https://github.com/theupdateframework/tuf/issues/1196

Lastly, test_slow_retrieval.py and slow_retrieval.py were refactored.
Until now, slow_retrieval.py couldn't use the TestServerProcess class
from utils.py for a port generation because of a bug related to
httpd.handle_request().
Now, when we use httpd.serve_forever() we can refactor both of those
files and fully conform with TestServerProcess.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Add tests for tests/utils.py

We want to make sure that server are successfully started in
the common use cases and that the new port generation works.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

jku commented on tests/utils.py:
```diff
@@ -154,69 +161,169 @@ class TestServerProcess():
+
+
+
+  def _log_queue_worker(self, stream, line_queue):
```
If the worker function is an object method (which should be fine), I think it would be good to mark it `@staticmethod `to clearly show that this function should never touch `self` state in any way, except via line_queue.

jku commented on tests/utils.py:
```diff
@@ -154,69 +161,169 @@ class TestServerProcess():
+    If it's found, self.port is set and returns True.
+    Otherwise, returns False.
+    TestServerProcessError is throwned if the process has exited
+    or if it has failed to bind.
```
```suggestion
    Raises TestServerProcessError if the process has exited.
```

jku commented on tests/utils.py:
```diff
@@ -154,69 +161,169 @@ class TestServerProcess():
+        raise TestServerProcessError(self.server + ' failed to bind!')
+    except queue.Empty:
+      self.__logger.info("No log lines are found in the Queue!")
+      return False
```
Since it's not called in a loop anymore, this function does not need a return value, I think. If you raise TimeoutError right here, change the method name to something like `wait_for_port()` and update the docstring, the boolean is no longer needed

Commit:
Make "bind successful" the first server message

The current implementation for server startup in TestServerProcess
relies on the fact that "bind successful..." is the first message
sent by the server process.
Make sure that this is true and leave a comment about this.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

jku commented on tests/utils.py:
```diff
@@ -154,69 +161,169 @@ class TestServerProcess():
+          # This means no more lines will be sent in the queue.
+          break
+        else:
+          self.__logged_messages.append(line)
```
```suggestion
        if len(line) > 0:
          self.__logged_messages.append(line)
```
Just a suggestion to make it easier to read: the break should not be required, next get() raises anyway

jku commented on tests/utils.py:
```diff
@@ -154,69 +161,170 @@ class TestServerProcess():
+    if len(self.__logged_messages) > 0:
       title = "Test server (" + self.server + ") output:"
-      message = [title] + log_message.splitlines()
+      message = [title] + self.__logged_messages
       self.__logger.info('\n| '.join(message))
```
I've not noticed `rstrip()` anywhere (and readline() results include '\n') so I expect this leads to double-newlines... Can you confirm that multiline logs still look good ?

MVrachev replied:
Yes, you were right:
![image](https://user-images.githubusercontent.com/16246778/98274541-36390d80-1f9c-11eb-8ba3-bd0230f123a2.png)

I will remove the `\n` from the ` self.__logger.info` call.

A good catch!

jku reviewed (commented):
It's been a long process (and that's not your fault) but I think we're about there now: this looks good. I have a couple of suggestions on the newest code (new compared to PR 1192): None of them show stoppers but please take a look.

Would be nice to get some more green CI runs in any case to confirm that this improves reliability like we think it will.

jku commented on tests/utils.py:
```diff
@@ -154,69 +161,170 @@ class TestServerProcess():
+        # Check if more lines are logged.
         self.flush_log()
-        raise e
+        raise TestServerProcessError(self.server + ' failed to bind!')
```
```suggestion
        raise TestServerProcessError(self.server + ' did not print port message as first stdout line as expected')
```
this might be fairer for the future testserver writer whose server binds just fine but does not print what we expect.

MVrachev commented on tests/test_multiple_repositories_integration.py:
```diff
@@ -134,20 +126,18 @@ def setUp(self):
 
     # Creates a subprocess running server and uses temp file for logging.
```
I have to remove all occurrences of `temp file` in the comments when creating the server subprocess...

Commit:
Tests: Remove "temp file" from comments

Update the comments to not mention the usage of temp file
for logging regarding the instances of the TestServerProcess class.

Also, remove one unused import.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I have updated the comments to not mention the usage of temp file
for logging regarding the instances of the TestServerProcess class.

That's why the number of changed files increased to 22, but most of the new changes are one-liners.

jku reviewed (approved):
LGTM, thanks

MVrachev commented:
I decided to rebase on top of the latest changes in `develop` branch.
I used that opportunity to squash some of my commits.

PS: Running the tests locally with `python aggregate_tests.py` on my Ubuntu 18.04 VM takes only 25~30 seconds after those changes!

joshuagl commented on tests/test_updater.py:
```diff
@@ -1863,27 +1864,35 @@ def setUp(self):
+    # in map.json, except for url prefixes.
+    # For the url prefixes, we create subprocesses that run a server script.
+    # In server scripts we get a free port from the OS which is sent
+    # back to the father process.
```
Nit:
```suggestion
    # back to the parent process.
```

joshuagl commented on tests/utils.py:
```diff
@@ -154,69 +161,161 @@ class TestServerProcess():
     self.server = server
-    self.port = port or random.randint(30000, 45000)
     self.__logger = log
+    # Stores poped messages from the queue.
```
Nit:
```suggestion
    # Stores popped messages from the queue.
```

joshuagl commented on tests/utils.py:
```diff
@@ -154,69 +161,161 @@ class TestServerProcess():
+    while True:
+      # readline() is a blocking operation.
+      # decode to push a string in the queue instead of 8-bit bytes.
+      log_line = stream.readline().decode()
```
```suggestion
      log_line = stream.readline().decode('utf-8')
```
Wasn't the default value of 'utf-8' for encoding only added in Python 3? Throughout the rest of the code we consistently call `decode()` with an explicit encoding of 'utf-8' and that seems to be a fairly consistent pattern in my experience. Is that redundant?

MVrachev replied:
It seems that for python2 the default is `ascii`. 
It's better to specify the encoding while we have python2 code.
By the way: a good catch.

joshuagl reviewed (commented):
This looks like a really solid set of improvements. Nice work @MVrachev! I've added a few minor comments while we wait for Travis CI.

joshuagl reviewed (approved):
Looks great, thanks @MVrachev! Could you squash 9bf3faa into cc95abb?

joshuagl commented:
Close/re-open to prompt CI
